src/pm-conversations/pmConversationsModel.js

```diff
@@ -0,0 +1,158 @@
+/* @flow strict-local */
+import Immutable from 'immutable';
+import invariant from 'invariant';
+import {
+  ACCOUNT_SWITCH,
+  EVENT_NEW_MESSAGE,
+  LOGIN_SUCCESS,
+  LOGOUT,
+  MESSAGE_FETCH_COMPLETE,
+  REALM_INIT,
+} from '../actionConstants';
+
+import type { Action, Message, Outbox } from '../types';
+import { recipientsOfPrivateMessage } from '../utils/recipient';
+
+//
+//
+// Keys.
+//
+
+/** The key identifying a PM conversation in this data structure. */
+// User IDs, excluding self, sorted numerically, joined with commas.
+export opaque type PmConversationKey = string;
+
+/** PRIVATE.  Exported only for tests. */
+// Input must have the exact right (multi-)set of users.  Needn't be sorted.
+export function keyOfExactUsers(ids: number[]): PmConversationKey {
+  return ids.sort((a, b) => a - b).join(',');
+}
+
+// Input may contain self or not, and needn't be sorted.
+function keyOfUsers(ids: number[], ownUserId: number): PmConversationKey {
+  return keyOfExactUsers(ids.filter(id => id !== ownUserId));
+}
+
+// Input must indeed be a PM, else throws.
+function keyOfPrivateMessage(msg: Message | Outbox, ownUserId: number): PmConversationKey {
+  return keyOfUsers(recipientsOfPrivateMessage(msg).map(r => r.id), ownUserId);
+}
+
+/** The users in the conversation, other than self. */
+export function usersOfKey(key: PmConversationKey): number[] {
+  return key ? key.split(',').map(s => Number.parseInt(s, 10)) : [];
+}
+
+//
+//
+// State and reducer.
+//
+
+/**
+ * The list of recent PM conversations, plus data to efficiently maintain it.
+ *
+ * This gets initialized from the `recent_private_conversations` data
+ * structure in the `/register` response (aka our initial fetch), and then
+ * kept up to date as we learn about new or newly-fetched messages.
+ */
+// (Compare the webapp's implementation, in static/js/pm_conversations.js.)
+export type PmConversationsState = {|
+  // The latest message ID in each conversation.
+  map: Immutable.Map<PmConversationKey, number>,
+
+  // The keys of the map, sorted by latest message descending.
+  sorted: Immutable.List<PmConversationKey>,
+|};
+
+const initialState: PmConversationsState = { map: Immutable.Map(), sorted: Immutable.List() };
+
+// Insert the key at the proper place in the sorted list.
+//
+// Optimized, taking O(1) time, for the case where that place is the start,
+// because that's the common case for a new message.  May take O(n) time in
+// general.
+function insertSorted(sorted, map, key, msgId) {
+  const i = sorted.findIndex(k => {
+    const id = map.get(k);
+    invariant(id !== undefined, 'pm-conversations: key in sorted should be in map');
+    return id < msgId;
+  });
+
+  if (i === 0) {
+    return sorted.unshift(key);
```

Comment:
That's based on this bit from [the Immutable.List docs](https://immutable-js.github.io/immutable-js/docs/#/List/unshift):
> Lists are immutable and fully persistent with O(log32 N) gets and sets, and O(1) push and pop.
>
> Lists implement Deque, with efficient addition and removal from both the end (push, pop) and beginning (unshift, shift).

Reply:
Ah OK, got it. That doesn't get a mention at the entry for [`.unshift`](https://immutable-js.github.io/immutable-js/docs/#/List/unshift), and I guess I didn't look closely enough at that discussion at the top.

Reply:
Indeed, that'd be good to clarify in the Immutable docs :slightly_smiling_face:

Reply:
> That doesn't get a mention at the entry for [`.unshift`](https://immutable-js.github.io/immutable-js/docs/#/List/unshift), and I guess I didn't look closely enough at that discussion at the top.

Oh, one hint I just noticed there in the `.unshift` entry -- this doesn't change that it should be made clearer, but it could be a helpful thing to notice -- is that it says this:
> Note: `unshift` can be used in `withMutations`.

In general the things that can be used in `withMutations` are the operations that efficiently act on just a small bit of the collection, with cost like O(1) or O(log N), and the things that can't are those that act on potentially the whole collection or a large swath of it. (I think the idea is that if you're doing one of the latter inside `withMutations`, then you're getting no benefit from that and might as well do so outside of `withMutations`, and this serves as a warning in case you *thought* you were doing something nice and efficient.) So when you see that note on an Immutable method, that's a signal that that's a cheap local operation and not an expensive whole-collection operation.

Reply:
> > Note: unshift can be used in withMutations.

Huh, interesting! It makes sense, but I wouldn't have been looking for it if you hadn't mentioned it; I'll keep that in mind. 🙂